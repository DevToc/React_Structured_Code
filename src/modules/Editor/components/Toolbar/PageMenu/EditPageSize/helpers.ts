import { PaperType, PageSizeUnit } from 'types/paper.types';
import { PageSizePreset } from 'constants/paper';
import { MAX_NUMBER_INPUT, MIN_NUMBER_INPUT, PIXEL_TO_CM_RATIO, PIXEL_TO_INCHES_RATIO } from './config';

/**
 * Precisely round a giving numeric value to specific number of deciaml place
 *
 * @param value - numeric value
 * @param decimalPlaces - round to number of decimal places
 * @returns
 */
const roundNumber = (value: number, decimalPlaces = 2) => {
  const division = Math.pow(10, decimalPlaces);
  return Math.round((value + Number.EPSILON) * division) / division;
};

/**
 * Convert pixel value to inch
 *
 * @param value - numeric value
 * @returns
 */
const pixelToInch = (value: number): number => roundNumber(value / PIXEL_TO_INCHES_RATIO);

/**
 * Convert pixel value to cm
 *
 * @param value - numeric value
 * @returns
 */
const pixelToCm = (value: number): number => roundNumber(value / PIXEL_TO_CM_RATIO);

/**
 * Convert inch value to pixel
 *
 * @param value - numeric value
 * @returns
 */
const inchToPixel = (value: number): number => Math.round(roundNumber(value * PIXEL_TO_INCHES_RATIO));

/**
 * Convert cm value to pixel
 *
 * @param value - numeric vlaue
 * @returns
 */
const cmToPixel = (value: number): number => Math.round(roundNumber(value * PIXEL_TO_CM_RATIO));

/**
 * Convert giving value to pixel
 *
 * @param value - numeric value
 * @param unit - giving value's unit in [px, in, cm]
 * @returns
 */
const toPixels = (value: number, unit = PageSizeUnit.px) => {
  switch (unit) {
    case PageSizeUnit.cm:
      return cmToPixel(value);
    case PageSizeUnit.in:
      return inchToPixel(value);
    default:
      break;
  }

  return value;
};

/**
 * Match preset paper size base on giving width and height
 *
 * @param pageSize
 * @returns  The paper type if size matches
 */
const getPaperType = ({ widthPx, heightPx }: { widthPx: number; heightPx: number }): PaperType | undefined => {
  for (const [type, preset] of Object.entries(PageSizePreset)) {
    if (
      preset.widthPx === Math.min(widthPx as number, heightPx as number) &&
      preset.heightPx === Math.max(widthPx as number, heightPx as number)
    ) {
      return type as PaperType;
    }
  }
};

/**
 * Format and convert giving pixel value to specific unit value
 *
 * @param value - Value in pixels
 * @param unit - Conversion unit
 * @returns  A value with specific unit
 */
const format = (value: string | number, unit: PageSizeUnit): string => {
  switch (unit) {
    case PageSizeUnit.px:
      return value.toString();
    case PageSizeUnit.in:
      return pixelToInch(Number(value)).toString();
    case PageSizeUnit.cm:
      return pixelToCm(Number(value)).toString();
    default:
      break;
  }

  return '';
};

/**
 * Check whether the value is whithin the min and max input range
 *
 * @param value - A numeric value
 * @returns
 */
const validateSizeInput = (value: unknown): boolean => {
  const checkValue = Number(value);
  return !isNaN(checkValue) && checkValue > MIN_NUMBER_INPUT && checkValue < MAX_NUMBER_INPUT;
};

export { pixelToInch, inchToPixel, pixelToCm, cmToPixel, toPixels, getPaperType, format, validateSizeInput };
