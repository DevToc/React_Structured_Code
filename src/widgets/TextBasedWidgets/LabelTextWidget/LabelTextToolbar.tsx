import { ChangeEvent, useCallback } from 'react';
import { Flex, Tooltip, IconButton } from '@chakra-ui/react';
import { useTranslation } from 'react-i18next';

import { normalizeNumber } from 'utils/number';
import { useWidget } from 'widgets/sdk';
import { FontStyleOption, FontWeightOption, LabelTextWidgetData, TextDecorationOption } from './LabelTextWidget.types';
import { ToolbarNumberInput } from 'modules/common/components/Toolbar/ToolbarNumberInput';
import { ColorPicker } from 'modules/Editor/components/ColorPicker';
import { FontFamilySelect } from 'modules/common/components/FontFamilySelect';

import { ReactComponent as BoldIcon } from 'assets/icons/bold.svg';
import { ReactComponent as ItalicIcon } from 'assets/icons/italic.svg';
import { ReactComponent as UnderlineIcon } from 'assets/icons/underline.svg';
import { ToolbarDivider } from 'modules/common/components/Toolbar/ToolbarDivider';
import { FontSizeSelect } from 'modules/common/components/FontSizeSelect';
import { Input } from 'modules/common/components/Input/Input';

export const LabelTextToolbar = () => {
  const { hasFontSizeOption } = useWidget<LabelTextWidgetData>();

  return (
    <Flex gap='2' align='center' data-testid='metric-text-widget-toolbar'>
      <ValueInput />
      <ToolbarDivider />
      <FontColorOption />
      <FontFamilyOption />
      {/* TODO should remove this check when toolbar is updated */}
      {!!hasFontSizeOption && <FontSizeOption />}
      <ToolbarDivider />
      <FontStyleOptions />
    </Flex>
  );
};

const ValueInput = () => {
  const { value, isNumeric, suffix, updateWidget } = useWidget<LabelTextWidgetData>();

  // On change callback if the label is numeric
  const updateNumericValue = useCallback(
    (newValue: number) => {
      const newValueNormalized = `${normalizeNumber(0, 100, newValue)}`;
      if (value === newValueNormalized) return;

      updateWidget({ value: newValueNormalized });
    },
    [updateWidget, value],
  );

  // On change callback if the label is non-numeric
  const updateStringValue = useCallback(
    (e: ChangeEvent<HTMLInputElement>) => {
      updateWidget({ value: e.target.value || '' });
    },
    [updateWidget],
  );

  return isNumeric ? (
    <ToolbarNumberInput
      value={parseFloat(value) || 0}
      label={'Chart Value'}
      suffix={suffix}
      onChange={updateNumericValue}
    />
  ) : (
    // TODO update the styling when a non-numeric label is required - currently there are no widgets that use a non-numeric label text widget
    <Input placeholder={'Label'} value={value} onChange={updateStringValue} size={'xs'} />
  );
};

const FontFamilyOption = () => {
  const {
    style: { fontFamily },
    updateWidget,
  } = useWidget<LabelTextWidgetData>();

  const updateFontFamily = useCallback(
    (newFont: string) => {
      updateWidget({ style: { fontFamily: newFont } });
    },
    [updateWidget],
  );

  return <FontFamilySelect onChange={updateFontFamily} fontFamily={fontFamily} />;
};

const FontSizeOption = () => {
  const {
    style: { fontSize },
    updateWidget,
  } = useWidget<LabelTextWidgetData>();

  const updateFontSize = (e: React.FormEvent<HTMLSelectElement>) => {
    const target = e.target as HTMLSelectElement;
    const value = target.value;
    if (!value) return;

    const newSize = parseFloat(value);
    updateWidget({ heightPx: newSize, style: { fontSize: newSize } });
  };

  return <FontSizeSelect value={Math.floor(fontSize)} onChange={updateFontSize} />;
};

const FontColorOption = () => {
  const {
    style: { color },
    updateWidget,
  } = useWidget<LabelTextWidgetData>();
  const { t } = useTranslation('editor_widget_toolbar', { keyPrefix: 'labelTextToolbar', useSuspense: false });
  const label = t('fontColor.label');

  const updateFontColor = useCallback(
    (newColor: string) => {
      updateWidget({ style: { color: newColor } });
    },
    [updateWidget],
  );

  return (
    <ColorPicker
      iconStyle='text'
      color={color ?? ''}
      label={label}
      onChange={updateFontColor}
      showNoColorOption={false}
    />
  );
};

const FontStyleOptions = () => {
  const {
    style: { fontWeight, fontStyle, textDecoration },
    updateWidget,
  } = useWidget<LabelTextWidgetData>();
  const { t } = useTranslation('editor_widget_toolbar', { keyPrefix: 'labelTextToolbar', useSuspense: false });

  const boldAriaLabel = t('fontStyle.bold.ariaLabel');
  const italicAriaLabel = t('fontStyle.italic.ariaLabel');
  const underlineAriaLabel = t('fontStyle.underline.ariaLabel');

  const isBold = fontWeight === FontWeightOption.Bold;
  const isItalic = fontStyle === FontStyleOption.Italic;
  const isUnderline = textDecoration === TextDecorationOption.Underline;

  const onSetBold = useCallback(() => {
    updateWidget({
      style: { fontWeight: !isBold ? FontWeightOption.Bold : FontWeightOption.Normal },
    });
  }, [isBold, updateWidget]);

  const onSetItalic = useCallback(() => {
    updateWidget({
      style: { fontStyle: !isItalic ? FontStyleOption.Italic : FontStyleOption.Normal },
    });
  }, [updateWidget, isItalic]);

  const onSetUnderline = useCallback(() => {
    updateWidget({
      style: { textDecoration: !isUnderline ? TextDecorationOption.Underline : TextDecorationOption.None },
    });
  }, [updateWidget, isUnderline]);

  return (
    <>
      <Tooltip hasArrow placement='bottom' label='Bold' bg='black'>
        <IconButton
          isActive={isBold}
          onClick={onSetBold}
          size='sm'
          aria-label={boldAriaLabel}
          variant='icon-btn-toolbar'
          icon={<BoldIcon />}
        />
      </Tooltip>
      <Tooltip hasArrow placement='bottom' label='Italic' bg='black'>
        <IconButton
          isActive={isItalic}
          onClick={onSetItalic}
          size='sm'
          aria-label={italicAriaLabel}
          variant='icon-btn-toolbar'
          icon={<ItalicIcon />}
        />
      </Tooltip>
      <Tooltip hasArrow placement='bottom' label='Underline' bg='black'>
        <IconButton
          isActive={isUnderline}
          onClick={onSetUnderline}
          size='sm'
          aria-label={underlineAriaLabel}
          variant='icon-btn-toolbar'
          icon={<UnderlineIcon />}
        />
      </Tooltip>
    </>
  );
};
