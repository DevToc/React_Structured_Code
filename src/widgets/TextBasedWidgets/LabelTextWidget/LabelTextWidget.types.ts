import { Widget } from 'types/widget.types';

export enum FontWeightOption {
  Normal = 'normal',
  Bold = 'bold',
}

export enum TextDecorationOption {
  None = 'none',
  Underline = 'underline',
}

export enum FontStyleOption {
  Normal = 'normal',
  Italic = 'italic',
}

type LabelTextStyle = {
  fontWeight: FontWeightOption;
  fontStyle: FontStyleOption;
  textDecoration: TextDecorationOption;
  fontSize: number;
  fontFamily: string;
  color: string;
};

interface LabelTextWidgetData extends Widget {
  value: string;
  suffix?: string;

  style: LabelTextStyle;
  isNumeric: boolean;

  // TEMP - boolean to determine whether the font size can be updated via toolbar
  hasFontSizeOption?: boolean;
}

export type { LabelTextWidgetData, LabelTextStyle };
