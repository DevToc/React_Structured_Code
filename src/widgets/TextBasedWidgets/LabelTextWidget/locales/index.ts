import EN_LABEL_TEXT_TOOLBAR from './en/toolbar.json';
import ES_LABEL_TEXT_TOOLBAR from './es/toolbar.json';
import FR_LABEL_TEXT_TOOLBAR from './fr/toolbar.json';

export { EN_LABEL_TEXT_TOOLBAR, ES_LABEL_TEXT_TOOLBAR, FR_LABEL_TEXT_TOOLBAR };
