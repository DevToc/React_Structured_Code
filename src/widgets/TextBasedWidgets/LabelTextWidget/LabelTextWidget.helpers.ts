import { WidgetType } from 'types/widget.types';
import {
  FontStyleOption,
  FontWeightOption,
  LabelTextStyle,
  LabelTextWidgetData,
  TextDecorationOption,
} from './LabelTextWidget.types';
import { VERSION } from './LabelTextWidget.upgrade';

type DefaultDataOptions = {
  metricOptions?: Partial<LabelTextWidgetData>;
  styleOptions?: Partial<LabelTextStyle>;
};

export const generateDefaultData = (options?: DefaultDataOptions) => {
  return {
    widgetType: WidgetType.LabelText,
    widgetData: {
      version: VERSION,
      widthPx: 100,
      heightPx: 37.5,
      topPx: 0,
      leftPx: 0,
      rotateDeg: 0,

      // metric text data
      value: '50',
      suffix: '%',
      isNumeric: true,
      style: {
        fontWeight: FontWeightOption.Normal,
        fontStyle: FontStyleOption.Normal,
        textDecoration: TextDecorationOption.None,
        fontSize: 25,
        fontFamily: 'Inter',
        color: '#000',
        ...options?.styleOptions,
      },
      ...options?.metricOptions,
    },
  };
};
