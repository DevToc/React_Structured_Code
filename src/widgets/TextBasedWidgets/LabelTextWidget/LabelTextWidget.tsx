import { useCallback, useEffect, forwardRef, useRef } from 'react';
import { Text } from '@chakra-ui/react';
import { CustomOnResize, CustomOnResizeEnd } from 'modules/Editor/components/BoundingBox';
import { ReadOnlyWidgetBase, WidgetBase, WidgetToolbar, useWidget, useEditor } from 'widgets/sdk';
import { isSideHandle } from 'utils/boundingBox';
import { LabelTextToolbar } from './LabelTextToolbar';
import { LabelTextWidgetData } from './LabelTextWidget.types';

export const LabelTextWidget = () => {
  const {
    widthPx,
    value,
    style: { fontSize, fontFamily, fontStyle, fontWeight, textDecoration },
  } = useWidget<LabelTextWidgetData>();
  const { boundingBox } = useEditor();

  const textRef = useRef<HTMLParagraphElement | null>(null);

  useEffect(() => {
    boundingBox.updateRect();
  }, [value, fontFamily, fontWeight, fontSize, fontStyle, textDecoration, boundingBox]);

  // When scaling with group, scale font size
  const customOnResize = useCallback(
    ({ event, onResize, isGroup, isResponsiveGroup }: CustomOnResize) => {
      const { direction } = event;

      // for side handles in group don't resize the labels
      if (isGroup && isSideHandle(direction)) return;

      onResize(event);
      if (!isGroup && !isResponsiveGroup) return;

      const target = event.target;
      const widgetEl = textRef.current;
      if (!widgetEl) return;

      const ratio = target.clientWidth / widthPx;
      widgetEl.style.fontSize = `${fontSize * ratio}px`;
    },
    [fontSize, widthPx],
  );

  // Scale font size in widget data and reset styling set in onResize fn
  const customOnResizeEnd = useCallback(
    ({ event, onResizeEnd, isGroup }: CustomOnResizeEnd) => {
      const scaleFontSize = isGroup && !isSideHandle(event.lastEvent.direction);

      const target = event.target;
      const ratio = target.clientWidth / widthPx;
      const newFontSize = ratio * fontSize;

      onResizeEnd(event, scaleFontSize ? { style: { fontSize: newFontSize } } : {});

      // Wait for font size to update in widget data, then remove fontSize styling on text element
      // setTimeout is required to avoid BB glitching when resizing due to the font-size not being set yet
      setTimeout(() => {
        if (!isGroup) return;

        const widgetEl = textRef.current;
        if (widgetEl) {
          widgetEl.style.fontSize = '';
        }
      }, 0);
    },
    [fontSize, widthPx],
  );

  return (
    <WidgetBase onResize={customOnResize} onResizeEnd={customOnResizeEnd}>
      <WidgetToolbar>
        <LabelTextToolbar />
      </WidgetToolbar>
      <LabelText ref={textRef} />
    </WidgetBase>
  );
};

export const ReadOnlyLabelTextWidget = () => {
  return (
    <ReadOnlyWidgetBase>
      <LabelText />
    </ReadOnlyWidgetBase>
  );
};

const LabelText = forwardRef<HTMLParagraphElement, {}>((_, ref) => {
  const { widgetId, value, suffix, style } = useWidget<LabelTextWidgetData>();
  const { fontStyle, fontWeight, textDecoration, fontSize, fontFamily, color } = style;

  return (
    <Text
      id={`labeltext-${widgetId}`}
      fontWeight={fontWeight}
      fontSize={`${fontSize}px`}
      fontFamily={fontFamily}
      color={color}
      textDecoration={textDecoration}
      fontStyle={fontStyle}
      textAlign={'center'}
      lineHeight={1}
      ref={ref}
    >
      {value}
      {suffix}
    </Text>
  );
});
