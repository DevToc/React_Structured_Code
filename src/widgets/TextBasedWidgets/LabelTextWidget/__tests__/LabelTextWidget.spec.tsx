import { renderWidget } from 'utils/test-utils.test';
import { LabelTextWidget } from '../LabelTextWidget';
import { generateDefaultData } from '../LabelTextWidget.helpers';

describe('widgets/TextBasedWidgets/LabelTextWidget', () => {
  it('should render correctly', () => {
    const widgetId = '020.widget-1';
    const widget = { widgetId, ...generateDefaultData() };

    const { asFragment } = renderWidget(<LabelTextWidget />, { widget });
    expect(asFragment()).toMatchSnapshot();
  });
});
