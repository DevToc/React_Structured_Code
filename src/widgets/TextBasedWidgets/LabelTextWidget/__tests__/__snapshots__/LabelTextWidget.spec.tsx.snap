// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`widgets/TextBasedWidgets/LabelTextWidget should render correctly 1`] = `
<DocumentFragment>
  <div
    aria-describedby="020.widget-1-description"
    aria-label=""
    class="widget-container widget-selecto-target  css-1eee8vp"
    data-testid="widgetbase-020.widget-1"
    id="020.widget-1"
    style="display: block; position: absolute; top: 0px; left: 0px; width: 100px; height: 37.5px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <p
      class="chakra-text css-z2mbuv"
      id="labeltext-020.widget-1"
    >
      50%
    </p>
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="020.widget-1-description"
      />
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
